File: Problems/22. Generate Parentheses.py
"""
Problem Link: https://leetcode.com/problems/generate-parentheses/
"""


class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        def generateParenthesis(result, current, open, close, max):
            if (len(current) == 2 * max):
                result.append(current)
                return

            if (open < max):
                generateParenthesis(result, current + "(", open + 1, close, max)
            if (close < open):
                generateParenthesis(result, current + ")",
                                    open, close + 1, max)

        result = []
        generateParenthesis(result, "", 0, 0, n)
        return result

File: Problems/22. Generate Parentheses.java
/*
Problem Link: https://leetcode.com/problems/generate-parentheses/
*/

import java.util.*;

class Solution {
    public void helper(List<String> ans, String curr, int open, int close, int max) {
        if (curr.length() == max * 2) {
            ans.add(curr);
            return;
        }

        if (open < max) {
            helper(ans, curr + "(", open + 1, close, max);
        }
        if (close < open) {
            helper(ans, curr + ")", open, close + 1, max);
        }
    }

    public List<String> generateParenthesis(int n) {
        List<String> ans = new ArrayList<>();
        helper(ans, "", 0, 0, n);
        return ans;
    }
}